Automattic/wp-calypso
# WIP: front end email password validation improvements

Related to 8968-gh-Automattic/dotcom-forge

## Proposed Changes

- Adds handling for back end password validation via a mock API call
- Adds validation checks to confirm that predictable terms like the domain name and mailbox name are not included in the password
- Updates email form validation to allow for and display multiple validation messages at a time
- Addresses a bug in the existing invalid character matching pattern
- Adds email addresses beginning with a hyphen to the invalid email address format validation

## Why are these changes being made?

- Currently the email password field does some light validation, covering things like password length, leading and trailing spaces, and forbidden characters (we only allow ASCII). We don't currently check for common or predictable passwords. The API call (endpoint is still a WIP) and checks for the domain and mailbox name will help prevent the use of weak/predictable passwords for new customer mailboxes
- There are places on the email form that could potentially have more than one error at a time (for example, a password could have a leading space, a trailing space, and contain the name on the mailbox). Currently we only show the first error we find, based on the order our validators are called. To make things easier for the user, we're now displaying any errors related to a given field at once, so they can be addressed together instead of one at a time. (see p1725450050739169/1725409874.123499-slack-C06ELHR6L9J for more context/discussion)
- A bug in the existing regex for invalid characters caused a `.` character at the end of the mailbox name to render an invalid character error, stating that only certain characters (including a `.`) are allowed. Adding more alphanumeric characters after the `.` cleared the error. This is pretty confusing as the mailbox name "firstname.@example.com" would be invalid despite not using any disallowed characters, but "firstname.lastname@example.com" would be fine. This happened because in addition to defining the allowed characters in the mailbox name, the regex pattern was also enforcing the order in which they could appear, requiring the `.` to be surrounded on either side by strings of other valid characters. While it's true that a trailing (or leading) `.` makes an invalid mailbox name, the validation error we showed was inaccurate. Instead, the regex now just defines the set of valid characters, and lets a different validator check the format of the requested email address, which gives us accurate errors.
- The existing validation allows for interesting use of both leading and trailing hyphens and underscores. After testing with Titan (I'll look at Google Workspace tomorrow), the only version that's actually a problem is a leading hyphen. All other uses of hyphens and underscores validate and provision successfully. A leading hyphen validates in the form, but fails during checkout due to some back end validation. Leading hyphens are also blocked by back end validation when adding new mailboxes to an existing Titan account. Our current email address format validation does not prevent this on our form, so I've added a check for leading hyphens to make the form consistent and predictable.

## ToDo
- I have a couple of code comments I need to remove, they're notes I want to add to this PR before its reviewed, please ignore them for now
- Test leading/trailing hyphens and underscores in Google Workspace accounts and confirm they behave the same (either due to the vendor or our own back end validation) and make sure this PR satisfies the needs of both products)

## Testing Instructions

- Select a site with a custom domain but no email upgrade and navigate to My Home >Upgrades > Email
- In the Titan email set up form, test the available fields
- Mailbox name:
  - leading hyphens aren't allowed, and prompts you to enter a valid email address
  - a leading/trailing dot isn't allowed, and prompts you to enter a valid email address, but does NOT trigger an invalid character warning
  - using characters other than letters, numbers, hyphens, dots, or underscores triggers an invalid character warning
 - Password:
   - Confirm that short passwords (under 10 characters), leading or trailing spaces, and non-ASCII characters still generate validation errors
   - Simulate a mock API validation by entering the password `password123`. These should both trigger "This password is too common" errors (a temporary value, the real error will probably come from the API point in the future)
   - Add the domain name of the site you're working with in the password field and confirm it triggers a validation error
   - Add a value into the mailbox name, and then put that same value into the password. Confirm this also invalidates the field
   - Finally, test any combination of the above password errors, and confirm that multiple errors render as expected

Required field handling
- In the password field, clear out any text you've entered
- Confirm you get the required field error, but NOT a "password too short" error. When a required field is empty, that should be the only error displayed.
